File: Problems/238. Product of Array Except Self/solution.js
/**
 * @param {number[]} nums
 * @return {number[]}
 */
 const productExceptSelf = nums => {
  let totalProduct = 1;
  let productWithoutZero = 1;
  let onlyZeros = true;
  let seenZero = false;
  for (let i = 0; i < nums.length; i++) {
      if (nums[i] === 0) {
          if (seenZero) {
              productWithoutZero *= 0;
          } else {
              seenZero = true;
          }
      } else {
          productWithoutZero *= nums[i];
          onlyZeros = false;
      }
      totalProduct *= nums[i];
  }
  
  const ans = [];
  for (let i = 0; i < nums.length; i++) {
      if (nums[i] === 0) {
          ans.push(onlyZeros ? 0 : productWithoutZero);
      } else {
          ans.push(totalProduct / nums[i]);   
      }
  }
  
  return ans;
};

// Time complexity: O(n)
// Space complexity: O(1), without taking into consideration the ans array.
// Pattern: Arrays
// Time used: Didn't count it but roughly 35-40min

File: Problems/238. Product of Array Except Self/solution.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        ans = []
        
        prefix = 1
        for i in range(len(nums)):
            ans.append(prefix)
            prefix *= nums[i]
            
        suffix = 1
        for i in range(len(nums) - 1, -1, -1):
            ans[i] *= suffix
            suffix *= nums[i]
            
        return ans

# Time complexity: O(n)
# Space complexity: O(1), because we're not taking into account the output array. 
# Pattern: Arrays        
# Time used before looking at solution: 15min